crates/ruma-identifiers/src/opaque_ids.rs

```diff
@@ -20,6 +22,21 @@ impl DeviceId {
     }
 }
 
+impl From<&DeviceId> for Rc<DeviceId> {
+    /// Convert DeviceId to Rc<DeviceId>
+    fn from(s: &DeviceId) -> Rc<DeviceId> {
+        let rc = Rc::<str>::from(s.as_str());
+        unsafe { Rc::from_raw(Rc::into_raw(rc) as *const DeviceId) }
+    }
+}
+impl From<&DeviceId> for Arc<DeviceId> {
+    /// Convert DeviceId to Arc<DeviceId>
+    fn from(s: &DeviceId) -> Arc<DeviceId> {
+        let arc = Arc::<str>::from(s.as_str());
+        unsafe { Arc::from_raw(Arc::into_raw(arc) as *const DeviceId) }
+    }
+}
```

Comment:
As mentioned in #615, this should be part of the `opaque_identifier!` macro, i.e. please add these after [the Box one](https://github.com/ruma/ruma/blob/c8f780dbbbc901980499b4c79e167253a4d2282f/crates/ruma-identifiers/src/macros.rs#L227-L231) and use `$id` instead of `DeviceId` (plus qualified paths for `Rc` / `Arc`).